ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp

```diff
@@ -345,10 +346,23 @@ static int HandleVnodeOperation(
         
         // Call vn_getpath first when the cache is hottest to increase the chances
         // of successfully getting the path
-        if (0 == vn_getpath(currentVnode, vnodePathBuffer, &vnodePathLength))
+        int error = vn_getpath(currentVnode, vnodePathBuffer, &vnodePathLength);
+        if (0 == error)
         {
             vnodePath = vnodePathBuffer;
         }
+        else
+        {
+            const char* name = vnode_getname(currentVnode);
+            mount_t mount = vnode_mount(currentVnode);
+            vfsstatfs* vfsStat = mount != nullptr ? vfs_statfs(mount) : nullptr;
+
+            KextLog_Error("HandleVnodeOperation: vn_getpath failed for vnode %p, error = %d, name '%s', recycled: %s, on mount point mounted at '%s'", currentVnode, error, name ?: "[NULL]", vnode_isrecycled(currentVnode) ? "yes" : "no", vfsStat ? vfsStat->f_mntonname : "[NULL]");
+            if (name != nullptr)
+            {
+                vnode_putname(name);
```

Comment:
For my own info, what happens in this case.  I see we call vnode_putname(name) here, what would be difference between this and going thru the normal flow.

Reply:
Memory for vnode names is reference counted (and pooled, I believe?), so if you successfully call `vnode_getname()` as we do 7 lines up, you need to match it with a call to `vnode_putname()`. Not to be confused with `vnode_get()` and `vnode_put()` which operate on the vnode object itself, not the name.

Reply:
Did you consider a RAII pattern to ensure we release the reference on `name`? I like letting the language ensure resource lifetime is tied to object lifetime, and it's a bit of extra safety if anyone introduces a new exit point when that reference is held.

Reply:
@jrbriggs I like to use RAII when possible as well.

@pmj what are your thoughts in general when it comes to using the RAII pattern in kext code?

Reply:
> what are your thoughts in general when it comes to using the RAII pattern in kext code?

I'm not generally opposed to it at all; I'm definitely a fan of getting the compiler to verify correctness as much as possible. I'd consider 2 major and 1 minor argument against wrapping this and some of the other resource releasing APIs, however:

1. The system APIs themselves don't return RAII types, so you need to either wrap all the kernel APIs with RAII-ified versions (and enforce using only them), *or* you need to RAII-ify correctly at each call site, which can itself become a potential hazard. `vnode_getname()` for example returns a `const char*`; this doesn't mean that every `const char*` can or should be wrapped in the same RAII class. For more complex APIs, such as those dealing with `vnode_t` themselves or `OSObject`, you have to be aware of whether a returned reference is retained or not (in the vnode case, the distinction between iocount and refcount makes it even more complicated).
2. Some things are best left explicit, particularly when interacting with locks/mutexes. There are certain APIs where it's very much relevant whether or not they're called inside or outside a held lock; when calling them explicitly, this intention is typically quite visible in the code. When they're called implicitly in a destructor, you have to infer the intention based on the scope of the RAII variable; sometimes your intention conflicts with C++ variable scoping rules, so you end up having to work around the scoping and perform the call explicitly in some way or another after all. If you're having to resort to that sort of thing, I'd consider that making things worse, not better.
3. *(This argument carries less weight.)* The RAII wrappers themselves need to be maintained and could themselves be sources of errors.

Item 2 could be counteracted by a kind of "RAII lite" where resource acquisition & release remain explicit, but are checked at runtime by assertions in the destructor - a.k.a. a hack for missing support for linear types in C++.

Reply:
@pmj Thanks for the detailed explaination on vnode_putname and detailed thoughts!

You mentioned the get/put pairing is for reference counting.  Out of curiosity what would be the downfall if we forget to do the put?  Would it solely be a memory leak or could it be more destructive.  Since theses are vnodes I'm guessing just a memory leak would be pretty dangerous, just wondering what other consequences could happen for my own understanding.  

Finally would a wrapper function that pairs the get/put, be a simpler alternative to RAII in this case.  Just thinking it would be a more lightweight.  Something like:
```
char* GetName(vnode)
{
  const char* name = vnode_getname(vnode); 
  if (name != nullptr) 
  { 
     vnode_putname(name); 
  }
  return name;
}
```

Reply:
> You mentioned the get/put pairing is for reference counting. Out of curiosity what would be the downfall if we forget to do the put? Would it solely be a memory leak or could it be more destructive. Since theses are vnodes I'm guessing just a memory leak would be pretty dangerous, just wondering what other consequences could happen for my own understanding.

I had to look it up in the source (bsd/vfs/vfs_cache.c in xnu), but it looks like it's a fairly straightforward memory leak. Obviously a memory leak in the kernel is worse than in user space because it's permanent until reboot, and it's wired memory, so can never be swapped out. Leaking a vnode itself (as opposed to its name) would arguably be worse, as we found out in #499.

> Finally would a wrapper function that pairs the get/put, be a simpler alternative to RAII in this case. Just thinking it would be a more lightweight. Something like:
> 
> ```
> char* GetName(vnode)
> {
>   const char* name = vnode_getname(vnode); 
>   if (name != nullptr) 
>   { 
>      vnode_putname(name); 
>   }
>   return name;
> }
> ```

It's a nice idea, but it's unfortunately not safe to dereference that pointer after you've relinquished your last refcount with `vnode_putname()`. For example, the vnode could be renamed or deleted by another thread while we're executing this code, which could cause the refcount on the name to drop to 0.

Reply:
Thanks @pmj that makes sense!  I'm learning so much on these PRs :-).

I do think a pattern that enforces get/put are always together would a good idea for safety.

Reply:
> I do think a pattern that enforces get/put are always together would be a good idea for safety.

I've now prototyped this [here](https://github.com/pmj/VFSForGit/commit/abd81ec931f0e636d5f17ed50d6b8341d1dcd28c).

I've not added it to this PR as I honestly don't think it's a sufficient improvement. You still have to remember to use the wrapper, you need to be aware of the wrapper's semantics (construction *transfers* ownership) and none of the instances where the code uses `vnode_getname()` is so complicated that the `vnode_putname()` call is likely to be missed out on some codepath or another.

Based on the observation that I've only used `vnode_getname()` for logging detailed vnode properties so far, I've gone in the opposite direction in the latest commit (524a78f) added to this PR. I've simply created logging wrapper macros along the lines of our existing `KextLog_File*` macros, but which output those extra details about the vnode. This cleans up the error site and moves the contentious `vnode_getname()/vnode_putname()` pair into a helper function where it's hopefully out of harm's way. (N.B. I still need to *test* this latest commit but barring any stupid typos it should work.)